app/models/detector/suggested_resource.rb

```diff
@@ -0,0 +1,57 @@
+# frozen_string_literal: true
+
+# == Schema Information
+#
+# Table name: detector_suggested_resources
+#
+#  id          :integer          not null, primary key
+#  title       :string
+#  url         :string
+#  phrase      :string
+#  fingerprint :string
+#  created_at  :datetime         not null
+#  updated_at  :datetime         not null
+#
+
+require 'stringex/core_ext'
+
+module Detector
+  # Detector::SuggestedResource stores custom hints that we want to send to the
+  # user in response to specific strings. For example, a search for "web of
+  # science" should be met with our custom login link to Web of Science via MIT.
+  class SuggestedResource < ApplicationRecord
+    before_save :update_fingerprint
+
+    def update_fingerprint
+      self.fingerprint = calculate_fingerprint
+    end
+
+    # This implements the OpenRefine fingerprinting algorithm. See
+    # https://openrefine.org/docs/technical-reference/clustering-in-depth#fingerprint
+    def calculate_fingerprint
+      temp = phrase
```

Comment:
Non-blocking: `temp` is not a super useful name for a variable. As we are working with modifying phrases, maybe something like `modifified_phrase` would help us out.

You also use a bunch of `!` operators which change the underlying string rather than creating a copy so this might not be required at all (I believe phrase.strip! is the same as temp = phrase; temp.strip!, etc)

Reply:
In general, I'm not going to defend the variable names in this method. I'm not sure at all about just applying these steps to `phrase`, though - it feels like that would alter the information we got from staff, and which they'll be looking for when they try to maintain these records.

Reply:
Oh, I don't particularly love the bang methods at all here. I'm just saying if you are going to use them, go nuts with them because you have determined it is safe to do so. They aren't necessary at all if you are assigning the output to a variable. Their entire point is to change the thing _in place_ unsafely.

From the docs:
```
The bang methods (! at the end of the method name) are called and executed just like any other method. However, by convention, a method with an exclamation point or bang is considered dangerous. In Ruby's core library the dangerous method implies that when a method ends with a bang (!), it indicates that unlike its non-bang equivalent, permanently modifies its receiver. Almost always, the Ruby core library will have a non-bang counterpart (method name which does NOT end with !) of every bang method (method name which does end with !) that does not modify the receiver. This convention is typically true for the Ruby core library but may or may not hold true for other Ruby libraries.
```

So it is entirely possible you _are_ modifying `phrase` but just didn't know it but using the bangs.

```ruby
irb(main):038:0> x = "HALLO"
=> "HALLO"
irb(main):039:0> y = x.downcase!
=> "hallo"
irb(main):040:0> x
=> "hallo"

```

Reply:
Essentially, if you want to safely use bangs here you should pass a string in as an argument rather than using the `phrase` attribute of the model.